src/client/Client.cc

```diff
@@ -2573,53 +2573,53 @@ bool Client::ms_dispatch2(const MessageRef &m)
   switch (m->get_type()) {
     // mounting and mds sessions
   case CEPH_MSG_MDS_MAP:
-    handle_mds_map(MMDSMap::msgref_cast(m));
+    handle_mds_map(ref_cast<MMDSMap>(std::move(m)));
```

Comment:
This is converting a const-lvalue to a non-const (?) rvalue? That doesn't seem safe? I'm surprised that even compiles.

Reply:
why do you think `m` is a const-lvalue? it is `const MessageRef &`. which means we cannot change `m`. but the raw pointer in it is mutable.

Reply:
I understand that `*m` is mutable but `m` should not be, right? i.e. `decltype(std::move(m))` would be `const MessageRef&&`? That doesn't seem to make sense.

Reply:
To answer my own question, I think this compiles because the compiler is choosing `boost::intrusive_ptr<T> ref_cast(const boost::intrusive_ptr<U>& r)` by creating a temporary from the const-rvalue. What do you think kefu?

Reply:
If that's the case, then the `std::move` has no useful effect. Let's remove it?

Reply:
Ahh, you are right! Will do.